src/k8s/pkg/k8sd/app/cluster_util.go

```diff
@@ -0,0 +1,130 @@
+package app
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"path"
+
+	"github.com/canonical/k8s/pkg/k8sd/api/impl"
+	"github.com/canonical/k8s/pkg/k8sd/pki"
+	"github.com/canonical/k8s/pkg/k8sd/setup"
+	"github.com/canonical/k8s/pkg/k8sd/types"
+	"github.com/canonical/k8s/pkg/snap"
+	snaputil "github.com/canonical/k8s/pkg/snap/util"
+	"github.com/canonical/k8s/pkg/utils/k8s"
+	"github.com/canonical/microcluster/state"
+)
+
+func setupDatastoreCertificates(snap snap.Snap, cfg types.ClusterConfig, nodeName string, allowSelfSignedCA bool) (error, *pki.K8sDqlitePKI, *pki.ExternalDatastorePKI) {
```

Comment:
this is not a useful function, as the semantics are blurred. first of all, the error ought to be the last return value. then, we return two distinct pkis, and we don't define when either will be nil.

this leads me to think that perhaps we should avoid this completely. if these are two separate code-paths in bootstrap and join, then let's treat them as such and not group them like this.

Reply:
I was beginning to think the same, this section looks the same on the first glance but actually isn't.